File: week06/7/[241007](Codility)(Stack,Queue)_Brackets.py
"""문제 : Brackets
A string S consisting of N characters is considered to be properly nested if any of the following conditions is true:
    S is empty;
    S has the form "(U)" or "[U]" or "{U}" where U is a properly nested string;
    S has the form "VW" where V and W are properly nested strings.
    For example, the string "{[()()]}" is properly nested but "([)()]" is not.
Write a function:
    def solution(S)
that, given a string S consisting of N characters, returns 1 if S is properly nested and 0 otherwise.

For example, given S = "{[()()]}", the function should return 1 and given S = "([)()]", the function should return 0, as explained above.

Write an efficient algorithm for the following assumptions:

N is an integer within the range [0..200,000];
string S is made only of the following characters: '(', '{', '[', ']', '}' and/or ')'.
"""

"""문제풀이
    1. stack 방식으로 풀면 된다. 
        가장 최근에 열린 bracket이 닫혀야 그 다음 닫기가 가능하다.
        bracket이 닫히려면, 그에 상응한 bracket이 열린상태여야 한다.
    2. open : (,  [, {
        close : ), ], }
    3. S를 순회하면서 새로운 리스트에 넣는다. 최종적으로 list에 아무것도 남아있지 않으면 1, 남아있으면 0이다.
        - { 또는 [ 또는 ( 이면 맨 뒤에 append.
        - }, ], )이 들어왔을 때는, list의 맨 뒤 요소가 상응하면 bracket이면 해당 bracket을 pop, 상응하지 않으면 그냥 넣기.
"""

def solution(S):
    check_l = []
    for s in S :
        if len(check_l) == 0 and s in ")]}":
            return 0 # 바로 wrong
        if s in "([{" :
            check_l.append(s)
        if s == ')' :
            if check_l[-1] == '(' :
                check_l.pop()
            else :
                check_l.append(s)
        if s == ']' :
            if check_l[-1] == '[' :
                check_l.pop()
            else :
                check_l.append(s)
        if s == '}' :
            if check_l[-1] == '{' :
                check_l.pop()
            else :
                check_l.append(s)
    
    return 1 if  len(check_l) == 0 else 0

if __name__ == "__main__":
    print(solution("{[()()]}")) #1
    print(solution("([)()]"))  #0

File: week06/7/[241007](Codility)(Stack,Queue)_Brackets.c
/* 문제 : Brackets
A string S consisting of N characters is considered to be properly nested if any of the following conditions is true:
    S is empty;
    S has the form "(U)" or "[U]" or "{U}" where U is a properly nested string;
    S has the form "VW" where V and W are properly nested strings.
    For example, the string "{[()()]}" is properly nested but "([)()]" is not.
Write a function:
    int solution(char *S);
    that, given a string S consisting of N characters, returns 1 if S is properly nested and 0 otherwise.

For example, given S = "{[()()]}", the function should return 1 and given S = "([)()]", the function should return 0, as explained above.

Write an efficient algorithm for the following assumptions:

N is an integer within the range [0..200,000];
string S is made only of the following characters: '(', '{', '[', ']', '}' and/or ')'.
*/

/* 문제풀이
    1. stack 원리 이용, list 대신 array로. 그리고 마지막 자리요소를 추적하기 위해서 int length 변수 관리.
    2. S를 순회하면서 별도의 리스트에 넣는다. 조건에 맞는 경우에만 pop. 나머지는 append. 
        - pop 조건 : close bracket 이면서 리스트의 마지막 요소가 상응하는 bracket일때.
        - 주의사항 : list에 아무것도 없는 상태에서 close bracket을 append해야하면 false.
    3. 2가 끝난후 vector의 길이가 0이면 true, 0이 아니면 false
*/

#include <stdio.h>
#include <stdlib.h>
#define max_length 200000

int solution(char *S){
    char check[max_length] ;
    int cur_len = 0;

    int i = 0;
    while (S[i]){
        if (S[i] == '('|| S[i] == '['|| S[i] == '{') check[cur_len++] = S[i];
        else { // (S[i] == ')'|| S[i] == '}'|| S[i] == ']')
            if (cur_len == 0) return 0;
            if (S[i] == ')') {
                if (check[cur_len-1] == '(') cur_len--;
                else check[cur_len++] = S[i];
            }
            if (S[i] == ']') {
                if (check[cur_len-1] == '[') cur_len--;
                else check[cur_len++] = S[i];
            } 
            if (S[i] == '}') {
                if (check[cur_len-1] == '{') cur_len--;
                else check[cur_len++] = S[i];
            }         
        }
        //현재 ch 확인이 끝나면
        i++;
    }
    // S 순회 종료 후 cur_len이 0인지 확인.
    return cur_len == 0? 1 : 0;
}

int main(){
    char S1[9] =  "{[()()]}";
    printf("%d\n", solution(S1)); 
    char S2[6] =  "([)()]";
    printf("%d\n", solution(S2)); 
    
    return 0;
}